core/trino-main/src/main/java/io/trino/execution/executor/TaskExecutor.java

```diff
@@ -854,13 +875,15 @@ public Set<TaskId> getStuckSplitTaskIds(Duration processingDurationThreshold, Pr
         private final Thread thread;
```

Comment:
RunningSplitInfo currently has this state "Thread" which can change what it's doing through a mechanism external to the class. Now the addition of `taskRunner` provides better (and sufficient) encapsulation here, so IMO we should remove the direct Thread access from here, and instead let all usecases go through to taskRunner, unless there's a particular reason.

Reply:
in fact, may be better to make this check while getting the stacktrace itself. i.e. only give a valid stacktrace if the the runner is indeed processing the split's code.

Reply:
Hey, taskRunner is a runnable and don't the thread object  saved. Are you suggesting getting the currentThread from TaskRunner?

Reply:
For the taskRunner, we still can't guarantee it is always doing processing for the Split.

Reply:
we do need the check to make sure that the `taskRunner` is processing the same split, that this PR is doing. my original suggestion was around simplifying it by not having to store Thread. But I tried it out, and seems that we can't get away without having to stash it somewhere. 

The current code that gets the stacktrace is potentially getting an "incorrect" stacktrace. And this PR is avoiding this mistake later through a callback when trying to kill the task. I was trying to think if we can stop making this mistake in the first place, by making this check when you return the stacktrace (check whether we're still on the same split). In that case, you wouldn't need to pass in the callback, but simply fail the task like your previous code was doing. 

```
RunningSplitInfo {

        ......

        public Optional<Thread> getThreadStackTrace()
        {
            return taskRunner.getThreadStackTrace(this, thread);
        }
}

TaskRunner {
        ........

        synchronized StackTrace isStillExecuting(RunningSplitInfo stuckSplitInfo, Thread thread) {
            if (stuckSplitInfo != null && currentThreadSplitInfo == stuckSplitInfo) {
                     return thread.getStackTrace();
            } 
        }
}

``` 

I guess the question is, is there an advantage to executing the `failTask` within the TaskRunner lock or just getting the right stacktrace is enough? (Haven't looked into whether failTask may be locking something else that'd potentially cause a deadlock. cc @losipiuk @arhimondr) 

Even though we have to store the thread, if possible, we can just expose the needed stacktrace from the splitinfo and leverage taskRunner for locking. However, having gone through the exercise, I don't feel super strongly about readability in one or the other (advtg will be that we avoid the callback)

Reply:
I am trying to delay acquiring the lock as much as I can(only when actually interrupting the split). This is to reduce the impact of lock. In the implementation, the lock is only acquired after the interrupter verified the wall time and stacktrace. This is to avoid the interrupting thread impact majority of taskRunner threads. I decided to implement this, because the JONI library is a corner cases and we shouldn't block unrelated(or checked) threads every few seconds/minutes just to verify JONI stacktrace. What's more, there is always a race condition when checking using the stacktrace(we can't freeze the stacktrace)


Deadlock means two threads hold each other's locks, Since currentRunningSplitInfo is newly added. I don't think failTask will acquire currentRunningSplitInfo and causing deadlock. Especially in the all 3 places that acquired currentRunningSplitInfo, the code didn't ask for more locks.

Reply:
> IMO we are trying to solve a small problem that could happen with an extremely tiny probability by introducing extra synchronization complexity that may backfire in a significantly more impactful way. It looks like killing a task under a lock is probably safe today (I’m not 100% sure though due to complexity of task management code), but it may backfire if the task execution internals ever change (e.g.: when failing a task is changed to be synchronous vs asynchronous)

+1. I do see value in fixing it, but worried from future changes/maintainability perspective. One option is that we can leave a really extensive comment about this potential corner case, and leave it at that. OR, instead of failing the task inside the lock - just log warning inside the lock, but fail outside of the lock.